### ISSUE TITLE
Investigate hanging grid components

### ISSUE THREAD

**Issue by aagonzales** [carbon-design-system/carbon#5892]
Component that would/could hang in the gutter of the grid:

- Accordion
- Button 
- Code snippet
- Content switcher
- Data table
- Modal
- Progress indicator 
- Structured list
- Tab
- Tile

**Comment by joshblack** [comment]
@aagonzales diving into this with @tw15egan, any chance we could get some examples for how these would be used in product to try so that we can try and build them out and see how things work? 👀

**Comment by joshblack** [comment]
Hey @aagonzales! 👋  Was hoping to work with @tw15egan this week on this, any chance you had examples handy?

**Comment by aagonzales** [comment]
@joshblack @tw15egan can you try looking at example on the website first. Most of the product example I have will be screen shots and not files with the grid turned on. After I finish my release work I can look at digging up and documenting some clear specs.

Good places to look in the mean time:
https://www.carbondesignsystem.com/guidelines/2x-grid/code#grid-modes
https://www.carbondesignsystem.com/components/accordion/usage#placement
https://carbon-website-git-fork-jeanservaas-update-component-doc-205a6c.carbon-design-system.vercel.app/components/button/usage#button-groups
https://www.carbondesignsystem.com/patterns/notification-pattern#banner

Also there are a lot of great Watson example work in this folder (i can't speak to its correct grid usage though atm): https://ibm.box.com/s/xtsoax4i4jsgy7fnevmm8hidbzqtcm2o

**Comment by joshblack** [comment]
@aagonzales when we were diving in last week we looked at the website and I think we found the following examples for hanging content:

<details>
<summary>
Button and tile
</summary>

![image](https://user-images.githubusercontent.com/3901764/85473519-f89d9000-b578-11ea-9e15-b8c5e0f70504.png)

</details>

<details>
<summary>
Tab and tile 
</summary>

![image](https://user-images.githubusercontent.com/3901764/85473552-081cd900-b579-11ea-8dc5-de5787a4e066.png)
</details>

<details>
<summary>
Tile
</summary>

![image](https://user-images.githubusercontent.com/3901764/85473577-123ed780-b579-11ea-9f63-07986adacc2d.png)
</details>

<details>
<summary>
Tab
</summary>

![image](https://user-images.githubusercontent.com/3901764/85473606-1bc83f80-b579-11ea-9349-23b2b1d37c57.png)
</details>

<details>
<summary>
Website tabs, tiles, and search
</summary>

![image](https://user-images.githubusercontent.com/3901764/85473648-2b478880-b579-11ea-9950-22731c150e0e.png)

</details>

Let me know if I missed any!

---

Based on the list above, it seemed like we were missing:

- Code snippet
- Content switcher
- Data table
- Modal
- Progress indicator
- Structured list

For the watson folder, are there any particular files I should be looking for? It's hard navigating through and seeing which files include a hanging moment or not.

---

For more info, a lot of the grid work is challenging since it involves aligning containers on columns and, unfortunately, anything inside a container can't really align to a column since there is no sub grid support on the web (Although we're hopeful that [this](https://caniuse.com/#search=subgrid) will be better supported in the future for CSS Grid).

This is where the request for examples come from so we can see if we can fit a component in this model or if it would need to be nested in a container 👀

**Comment by aagonzales** [comment]
**Accordion:** 
Redesigned from IBM Cloud Support
(oh the last tile is wrong in this example fyi)

![image](https://user-images.githubusercontent.com/11670886/85474624-947bcb80-b57a-11ea-9360-d4c5081440b3.png)

**Data table:**
From: WS_My-projects_HIFI_200302.sketch
![image](https://user-images.githubusercontent.com/11670886/85474993-0a803280-b57b-11ea-9128-928d84fb749d.png)

**Code snippets**
You can see at work hanging in the gutter on our website
![image](https://user-images.githubusercontent.com/11670886/85475203-621e9e00-b57b-11ea-811c-48b418164beb.png)

**Modal**
Theoretically it hangs but since its not grid aware it width is a percentage. So maybe modal doesn't actually apply to this.
![image](https://user-images.githubusercontent.com/11670886/85475304-837f8a00-b57b-11ea-84e0-d4f8c8fa5253.png)


I'll try and dig up some more. I think I can find some in Slack, i'll also see if Jeannie has some more example. Full product screen example in V10 correctly are hard to get. I'm always begging designers to share files with me, most of the time we just get screen shots.

**Comment by jeanservaas** [comment]
Here's some examples from product:

Tomorrow I will keep looking for these (I know they exist):

- content switcher
- Progress indicator
- structured list


## Baremetal provisioning 

**tile**

<img width="1312" alt="image" src="https://user-images.githubusercontent.com/43144260/85482879-c3ea0280-b591-11ea-81f1-0185a964807b.png">

**underline tabs**

<img width="1312" alt="image" src="https://user-images.githubusercontent.com/43144260/85482917-d6643c00-b591-11ea-9228-7f3c2acf069d.png">

**container tabs**

<img width="1315" alt="image" src="https://user-images.githubusercontent.com/43144260/85483147-4e326680-b592-11ea-87e8-cd891449d296.png">

**input fields**
This was an interesting example that we found in the wild, a lot of Cloud designers were using the right side hang to allow input fields to have a 16px gutter. We haven't necessarily outlawed it but we haven't shown it in any guidance yet either.

<img width="1325" alt="image" src="https://user-images.githubusercontent.com/43144260/85483347-9fdaf100-b592-11ea-8455-19bcff119aee.png">


## Cloud dashboard

**button**

<img width="1312" alt="image" src="https://user-images.githubusercontent.com/43144260/85483023-162b2380-b592-11ea-89eb-682fb6bc51c5.png">

## Cloud Direct Link Dedicated

**selectable tiles with labels**

<img width="879" alt="image" src="https://user-images.githubusercontent.com/43144260/85484604-27296400-b595-11ea-9f75-f2806e4b3cea.png">

**data table**

<img width="831" alt="image" src="https://user-images.githubusercontent.com/43144260/85484717-65bf1e80-b595-11ea-9efd-c77affce1403.png">

## Cloud Catalog from Voltron

**hanging filterable search**

<img width="999" alt="image" src="https://user-images.githubusercontent.com/43144260/85484832-98691700-b595-11ea-8660-bcfb4ff437ed.png">


## Security ISC 

**Threat landscape layout**
I'm not really noticing a lot of hanging data tables in product, even Security usually keeps them on the columns, but often puts data tables on tiles to create the effect of a hang.
<img width="681" alt="image" src="https://user-images.githubusercontent.com/43144260/85483705-5c34b700-b593-11ea-93aa-43b04ec65a07.png">

**Threat report layout**
Here I do notice a fluid button on a tile... we don't show this in our guidance but it exists out in product.

<img width="682" alt="image" src="https://user-images.githubusercontent.com/43144260/85483855-a0c05280-b593-11ea-8dea-58e3618e70bf.png">

## Gatsby

**Search**

This isn't technically a product example, but I'm sure this is also done in product... hanging the search

<img width="1250" alt="image" src="https://user-images.githubusercontent.com/43144260/85484486-eb8e9a00-b594-11ea-8e08-6bca05c0f30f.png">

**Comment by joshblack** [comment]
Hell yeah, thanks you two! This rocks 🎸

**Comment by alisonjoseph** [comment]
_TLDR: Trying to replicate all the grid modes in code is super messy and hard, and developers will likely get it wrong even if we give them ways to do it._

I started playing around in codesandbox with some ideas and seeing how things could be laid out. 
https://codesandbox.io/s/hanging-explorations-ludzt?file=/index.js

Trying to build these layouts is super confusing and tricky, since components aren't grid aware, any dev building will have to either manually hang something if using the wide/standard grid, or manually add padding back in for components that should not hang if using condensed or narrow modes. 

For designs using “narrow” you could just use the default/wide grid mode and add a hang class/prop manually. Alternatively you could also use the narrow prop per row or column to remove the left padding/margin, but then you run into the issue if there are other items in that row/column that should not hang. Which is why it might make sense to do it per component. 

Where things get really weird is condensed. A dev would likely only use this at the row level or column level, however you’d have to be careful not to put anything in that column/row that shouldn’t hang. 

_Note: our grid doesn't currently support column level modes/classnames, although could be added. Only supported at entire grid or row levels._ 

To manually "hang" components it would be on the developer to know which components should hang so they can add the correct class (or prop if we added that) to each of them. Could get messy real quick when building out an entire application. We've seen this ourselves when building out components for the gatsby theme and website, and that's with our designers who understand how the grid should work by our side working closely. 

Could possibly do something where if "narrow" was set at the grid level, then any components inside the grid that should hang, would hang, this would need a lot of testing and could get messy. Not sure there would be a logical way to do this with condensed. Condensed would need to stay at grid/row/column level, and again, then its on the developer to make sure they manually add padding back in for components that shouldn't hang if inside those grid classes.

**Comment by alisonjoseph** [comment]
### **Proposed next steps for Grid**

- Add support to nest grids and have columns calculate correctly (one level)
- Add column level support for grid modes `.bx--col--condensed` and `.bx--col--narrow`
- Add a hang prop for certain components that need to hang
- Overhaul dev docs for grid to explain the grid modes more clearly and how to use at column/row level and hang
- Create a gallery of example layouts/templates
